docs/proposal/20220929-ipam-support.md

```diff
@@ -0,0 +1,422 @@
+- Support for specific IPs for specific machines - a machine will receive a
+  random IP in the pool.
+
+## Proposal
+
+Following the contract specified by the IPAM integration proposal we plan to:
+
+- Add the fromPools field to the spec.network.devices section of the
+  VSphereMachine and VSphereMachineTemplate. The fromPools field will allow a
+  user to specify that the network device IPAddress should be claimed from an
+  IPPool.
+- If a VSphereMachine has been specified to claim an IP Address from an IPPool,
+  then CAPV will create an IPAddressClaim, which the IPAM provider will use to
+  allocate an IP Address. This IPAddress can then be used by the
+  VSphereMachine.
+
+### User Stories
+
+#### Story 1
+
+As a user, I want to be able to create a VSphereMachine where the IP Address
+for the machine comes from an IPAM Provider instead of DHCP.
+
+#### Story 2
+
+As a user I want to be able to allocate either IPv4 or IPv6 addresses for my
+VSphereMachine.
+
+#### Story 3
+
+As a user I want to be able to release my IP Address allocation when the
+VSphereMachine is deleted for future VSphereMachines to use.
+
+### Implementation Details/Notes/Constraints
+
+#### API Changes
+
+Adding `fromPools`(type []corev1.TypedLocalObjectReference) to the VSphereMachineTemplate
+
+```yaml
+apiVersion: infrastructure.cluster.x-k8s.io/v1alpha3
+kind: VSphereMachineTemplate
+metadata:
+  name: example
+  namespace: vsphere-site1
+spec:
+  template:
+    spec:
+      cloneMode: FullClone
+      numCPUs: 8
+      memoryMiB: 8192
+      diskGiB: 45
+      network:
+        devices:
+        - dhcp4: false
+          dhcp6: false
+          networkName: VM Network
+          fromPools:
+            # reference to the pool
+            - group: ipam.cluster.x-k8s.io/v1alpha1
+              kind: InClusterIPPool
+              name: testpool4
+            - group: ipam.cluster.x-k8s.io/v1alpha1
+              kind: InClusterIPPool
+              name: testpool6
+```
+
+Add `FromPools` to the NetworkDeviceSpec:
+
+```go
+// NetworkDeviceSpec defines the network configuration for a virtual machine's
+// network device.
+type NetworkDeviceSpec struct {
+ // FromPools is a list of references to the pools from which IP addresses should be claimed.
+ FromPools []corev1.TypedLocalObjectReference `json:"fromPools,omitempty"`
+}
+```
+
+#### VSphereVM Controller Changes
+
+1. If the FromPools field is set on a device then this controller will create a
+   new IPAddressClaim for each pool configured on each device. The
+   IPAddressClaim will have an annotation containing the mac address of the
+   device, which will be used to correlate the IPAddress to the correct device.
+   The controller will not continue the VM creation process until all
+   IPAddressClaims have been fulfilled.
+
+   **Note**: The name of the IP Address Claim will have an appended randomly
+   generated hash for uniqueness by use of the `generatedName` field

```

Comment:
So `<machine name>-<hash>`?

Reply:
This is outdated too. The claims will have `<vm-name>-<device-dx>-<pool-idx>` names, per your suggestion.